### PR TITLE
Issue #424 - Implemented delete view for floors

### PR DESCRIPTION
**Pull request by emilianodellacasa**
For issue #424:

- refactored test to support generic views;
- added link to floor list pages to delete floor;
- left puzzle to ask for confirmation before deleting;

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#493](https://github.com/timelesslounge/timelessis/issues/493) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#493](https://github.com/timelesslounge/timelessis/issues/493) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/5b8b4d69-5e51-459b-a192-80e46520493b); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: timeless/restaurants/floors/views.py [hunk 1]
```diff
@@ -55,12 +53,11 @@ def edit(id):
         )
 
 
-@BP.route("/delete", methods=["POST"])
-@auth.login_required
-def delete():
-    """ Delete floor with id """
-    flash("Delete not yet implemented")
-    return redirect(url_for("floor.list_floors"))
+class Delete(views.DeleteView):
```
Comment (amihaiemil) [inline timeless/restaurants/floors/views.py]: @emilianodellacasa It seems strange to me that Floors can be deleted... a restaurant's floors are part of the building, they will never be deleted in real life. Maybe deactivated/put on hold/closed temporarily. 

Can we rename Delete to Deactivate? And not remove the Floor, but set a flag on it. WDYT?

File: timeless/restaurants/floors/views.py [hunk 2]
```diff
@@ -17,11 +17,9 @@
...
  must save the data to the repository. The tests must include checking if the
  view screens were correctly built and if the data was saved to the repository.
-@todo #95:30min Implement delete floor view. Delete floor view is composed of
- two views: the first view must load the desired floor data onto the screen
- and show an button to delete the selected view, extending
- timeless/views.py::DetailView; second view must show the result message of
- floor deletion.
+@todo #424:30min The last step for deletion is to ask for confirmation before
+ actually deleting the object. When clicking on link on list floors pages a
+ javascript modal should appear asking user for confirmation
```
Comment (amihaiemil) [inline timeless/restaurants/floors/views.py]: @emilianodellacasa Is this a puzzle that should be in the backend? Besides, I guess it should be a generic solution for all delete operations, not just for floors.

File: timeless/restaurants/floors/views.py [hunk 3]
```diff
@@ -55,12 +53,11 @@ def edit(id):
...
 
-@BP.route("/delete", methods=["POST"])
-@auth.login_required
-def delete():
-    """ Delete floor with id """
-    flash("Delete not yet implemented")
-    return redirect(url_for("floor.list_floors"))
+class Delete(views.DeleteView):
+    """ Delete floor with id """  
+    decorators = (auth.login_required,)
```
Comment (amihaiemil) [inline timeless/restaurants/floors/views.py]: @emilianodellacasa Can you remove the ``,``?

File: tests/integration/it_floors_test.py [hunk 4]
```diff
@@ -65,5 +64,8 @@ def test_edit(client, auth):
 @pytest.mark.skip(reason="auth.login() is not yet implemented")
 def test_delete(client, auth):
     auth.login()
-    response = client.post("/floors/delete", data={"id": 1})
+    floor = factories.FloorFactory()
+    response = client.post(url_for("floor.delete", id=floor.id))
     assert response.headers["Location"] == "http://localhost/floors/"
+    assert response.status_code == HTTPStatus.FOUND
+    assert Floor.query.count() == 0
```
Comment (amihaiemil) [inline tests/integration/it_floors_test.py]: @emilianodellacasa Can you also add a "negative" test? What happens if the delete is called with no authenticated user? They should be redirected to login, right?

File: timeless/restaurants/floors/views.py (hunk 1 above)
Comment (emilianodellacasa) [reply timeless/restaurants/floors/views.py]: @amihaiemil you comment is correct, before when we had the same issue with other entities POs suggested that it should be possible to completely destroy data from database. Sometimes, you just made a mistake and deleting is the right fix. We could of course develop the mechanism you suggest, especially adding a flag marking the floor as deleted (and not showing in the list).

@zverovich @virt13  What do you guys think?

File: timeless/restaurants/floors/views.py (hunk 2 above)
Comment (emilianodellacasa) [reply timeless/restaurants/floors/views.py]: @amihaiemil You are right of course, I wanted this to be generic. Let me specify it better

File: timeless/restaurants/floors/views.py (hunk 3 above)
Comment (emilianodellacasa) [reply timeless/restaurants/floors/views.py]: @amihaiemil No, it's a list

File: tests/integration/it_floors_test.py (hunk 4 above)
Comment (emilianodellacasa) [reply tests/integration/it_floors_test.py]: @amihaiemil There was one, now I fixed it! Thanks for pointing it out

File: timeless/restaurants/floors/views.py (hunk 1 above)
Comment (zverovich) [reply timeless/restaurants/floors/views.py]: @emilianodellacasa At the moment, we need only the basic operations above the floors Add / Rename / Delete.
You also need to implement a confirmation when user tries to delete a floor that has at least one table. Something like this: "Do you really want to delete the floor? All tables belonging to this floor will be also deleted!”

Comment (emilianodellacasa) [reply timeless/restaurants/floors/views.py]: @zverovich Thanks a lot, noted!

**Comment by emilianodellacasa** [comment]
@amihaiemil See comment from PO and my last commits

**Comment by amihaiemil** [comment]
@emilianodellacasa looks nice, thanks! :)

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 17 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#493](https://github.com/timelesslounge/timelessis/issues/493) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by amihaiemil** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +6 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**